File: lists/1.go
package main

import "fmt"

// Node comment
type Node struct {
	val int
	next  *Node
}

var root = new(Node)

func addNode(node *Node, data int) int {

	if root == nil {
		node = &Node{data, nil}
		root = node
		return 0
	}

	for node.next != nil {
		node = node.next
	}

	node.next = &Node{data, nil}
	return 1
}

func recursivelookUpNode(t *Node, data int) bool {
	if root == nil {
		t := &Node{data, nil}
		root = t
		return false
	}

	if data == t.val {
		return true
	}

	if t.next == nil {
		return false
	}

	return recursivelookUpNode(t.next, data)
}

func iterativelookUpNode(t *Node, data int) bool {
	if root == nil {
		t := &Node{data, nil}
		root = t
		return false
	}

	for t.next != nil {
		if data == t.val {
			return true
		}
		t = t.next
	}

	return false
}

// func deleteFromHead() {
// fmt.Printf(" %d -->", node, Node)
// node = Node.next
// }

// func deleteFromTail(data int) int {

// }
// func deleteByIndex(data int) int {

// }
// func deleteByval(data int) int {

// }

func printList(node *Node) {
	if node == nil {
		fmt.Println(" Empty List")
		return
	}

	for node != nil {
		fmt.Printf(" %d -->", node.val)
		node = node.next
	}
	fmt.Println()
}

func size(node *Node) int {
	if node == nil {
		fmt.Println(" Empty List")
		return 0
	}

	size := 0

	for node != nil {
		size++
		node = node.next
	}
	return size
}

func main() {
	root = nil
	addNode(root, 1)
	addNode(root, 2)
	addNode(root, 3)
	addNode(root, 4)
	printList(root)
	fmt.Println("Is 78 availilable in single LL -> ", recursivelookUpNode(root, 78))
	fmt.Println("Is 3 availilable in single LL -> ", recursivelookUpNode(root, 3))
	fmt.Println("Is 78 availilable in single LL -> ", iterativelookUpNode(root, 78))
	fmt.Println("Is 3 availilable in single LL -> ", iterativelookUpNode(root, 3))
	fmt.Println("List size -> ", size(root))
}

File: lists/1.js
const LL1 = function () {
  this.head = null

  this.push = (el) => {
    const node = {
      value: el,
      next: null,
      prev: null
    }
    if (!this.head) {
      this.head = node
    } else {
      current = this.head
      while (current.next) {
        current = current.next
      }
      current.next = node
    }
  }

  this.deleteFromHead = () => {
    let current = this.head
    current = current.next
    return current
  }

  this.deleteFromTail = () => {
    let current = this.head
    while (current.next.next) {
      current = current.next
    }
    current.next = null
  }

  this.deleteByIndex = (index) => {
    let count = 1
    let current = this.head
    while (current.next) {
      ++count
      process.stdout.write(String(current.value))
      process.stdout.write('--> ')
      if (index == count && current.next.next) {
        current = current.next.next
      } else {
        current = current.next
      }
    }
    console.log('null')
  }

  this.deleteByValue = (el) => {
    let current = this.head
    while (current.next) {
      if (current.next.value == el && current.next.next) {
        current = current.next.next
      } else {
        current = current.next
      }
    }
  }

  this.printList = () => {
    if (!this.head) {
      process.stdout.write('empty list')
    } else {
      current = this.head
      while (current.next) {
        process.stdout.write(String(current.value))
        process.stdout.write('--> ')
        current = current.next
      }
      console.log('null')
    }
  }

  this.printListByValue = (el) => {
    if (!this.head) {
      process.stdout.write('empty list')
      // return
    } else {
      current = this.head
      while (current.next) {
        if (current.value == el) {
          break
        } else {
          current = current.next
        }
      }
      while (current.next) {
        process.stdout.write(String(current.value))
        process.stdout.write('--> ')
        current = current.next
      }
      console.log('null')
    }
  }
}

const ll1 = new LL1()

ll1.push(1)
ll1.push(2)
ll1.push(3)
ll1.push(4)
ll1.push(5)
ll1.push(6)
ll1.push(7)
ll1.push(8)
console.log('final linkedlist', ll1.printList())
const val = ll1.deleteFromHead()
console.log('current head after delte head node', val.value)

console.log('list after deleteFromHead -> ')
ll1.printListByValue(val.value)
console.log('list after deleteFromTail -> ')
ll1.deleteFromTail()
console.log('list after deleteByIndex -> ')
ll1.deleteByIndex(2)
console.log('list after deleteByValue -> ')
ll1.deleteByValue(6)
console.log('---------------')
